[Bugfix] Preserve tool call id/type/name in streaming finish chunk

amittell opened the pull request:
## Summary

When streaming tool calls, the finish chunk code in `serving_chat.py` overwrites the tool parser's `DeltaMessage` with a stripped-down version containing only `index` and `function.arguments`. This loses the `id`, `type`, and `function.name` fields.

This fix preserves those fields from the original `delta_message` when building the finish chunk.

## Issue

Fixes #31437

## What was wrong

```python
delta_message = DeltaMessage(
    tool_calls=[
        DeltaToolCall(
            index=index,
            function=DeltaFunctionCall(
                arguments=remaining_call
            ).model_dump(exclude_none=True),
        )
    ]
)
```

The `id`, `type`, and `function.name` fields were lost.

## The fix

```python
original_tc = delta_message.tool_calls[0]
original_fn = original_tc.function if original_tc else None
delta_message = DeltaMessage(
    tool_calls=[
        DeltaToolCall(
            index=index,
            id=original_tc.id if original_tc else None,
            type=original_tc.type if original_tc else None,
            function=DeltaFunctionCall(
                name=original_fn.name if original_fn else None,
                arguments=remaining_call,
            ),
        )
    ]
)
```

## Testing

Tested with GLM-4.7-AWQ using `--tool-call-parser glm47`. Before fix: ~20% of streaming tool call responses had `id=None`. After fix: 100% success rate.

## Related

- #16340 - Similar symptoms, different root cause
- #10781 - Mentions delta not being submitted correctly

amittell committed:
fix: preserve tool call id/type/name in streaming finish chunk

When building the finish chunk for streaming tool calls, the code was
creating a new DeltaMessage with only index and arguments, losing the
id, type, and function.name fields that the tool parser had set.

This breaks OpenAI-compatible clients that expect id to be present.

Fix by preserving these fields from the original delta_message.

Fixes #31437

Signed-off-by: amittell <mittell@me.com>

amittell committed:
fix: match tool call by index for multi-tool scenarios

Address review feedback: use next() to find the tool call matching
the current index instead of assuming tool_calls[0] is correct.

Signed-off-by: amittell <mittell@me.com>

amittell committed:
style: apply ruff formatting

Signed-off-by: amittell <mittell@me.com>

chaunceyjiang commented on vllm/entrypoints/openai/serving_chat.py:
```diff
@@ -1208,13 +1208,28 @@ async def chat_completion_stream_generator(
+                                ),
+                                None,
+                            )
+                            original_fn = original_tc.function if original_tc else None
```
Thanks~ @amittell 

I noticed that #31451 seems to be addressing the same issue.

What do you think about #31451?

amittell commented:
Thanks for the pointer. I looked at #31451 - it skips delta creation when `remaining_call` is empty or matches current args.

One difference: when `remaining_call` is non-empty and different from current args, #31451 still creates a stripped delta without id/type/name. My fix preserves those fields in all cases by matching on index.

Refactored into a helper function per your suggestion:

```python
@staticmethod
def _create_remaining_args_delta(
    delta_message: DeltaMessage,
    remaining_call: str,
    index: int,
) -> DeltaMessage:
    """
    Create a delta message for remaining tool arguments, preserving
    id/type/name from the original delta.
    """
    original_tc = next(
        (tc for tc in delta_message.tool_calls if tc.index == index),
        None,
    )
    original_fn = original_tc.function if original_tc else None
    return DeltaMessage(
        tool_calls=[
            DeltaToolCall(
                index=index,
                id=original_tc.id if original_tc else None,
                type=original_tc.type if original_tc else None,
                function=DeltaFunctionCall(
                    name=original_fn.name if original_fn else None,
                    arguments=remaining_call,
                ),
            )
        ]
    )
```

Tested with GLM-4.7-AWQ on a 4-node cluster - 100% success rate. I can push this refactor if it looks good.

amittell committed:
refactor: extract _create_remaining_args_delta helper function

Per reviewer feedback, extracted the tool call delta creation logic
into a static helper method for better readability and testability.

Signed-off-by: amittell <mittell@me.com>

amittell committed:
test: add unit tests for _create_remaining_args_delta

Add tests to verify the helper function correctly preserves
id/type/name fields when building the finish chunk for streaming
tool calls.

Tests cover:
- Basic preservation of id/type/name from original delta
- Correct matching by index with multiple tool calls
- Graceful handling when no matching tool call is found
- Handling when original tool call has no function

Signed-off-by: Alex Mittell <mittell@me.com>
Signed-off-by: amittell <mittell@me.com>

chaunceyjiang reviewed (commented):
Thanks~ @amittell
